lib/Runtime/Library/RegexHelper.cpp

```diff
@@ -792,8 +782,234 @@ namespace Js
+
+                RecyclableObject* resultObj = ExecResultToRecyclableObject(result);
+
+                results->Add(resultObj);
+
+                if (!global)
+                {
+                    break;
+                }
+
+                JavascriptString* matchStr = GetMatchStrFromResult(resultObj, scriptContext);
+                AdvanceLastIndex(thisObj, input, matchStr, unicode, scriptContext);
+            }
+
+            CompoundString::Builder<64 * sizeof(void *) / sizeof(wchar_t)> accumulatedResultBuilder(scriptContext);
+            CharCount inputLength = input->GetLength();
+            CharCount nextSourcePosition = 0;
+
+            size_t previousNumberOfCapturesToKeep = 0;
+            Var* captures = nullptr;
+
+            results->Map([&](int i, RecyclableObject* resultObj) {
+                int64 length = JavascriptConversion::ToLength(
+                    JavascriptOperators::GetProperty(resultObj, PropertyIds::length, scriptContext),
+                    scriptContext);
+                uint64 numberOfCaptures = (uint64) max(length - 1, (int64) 0);
+
+                JavascriptString* matchStr = GetMatchStrFromResult(resultObj, scriptContext);
+
+                int64 index = JavascriptConversion::ToLength(
+                    JavascriptOperators::GetProperty(resultObj, PropertyIds::index, scriptContext),
+                    scriptContext);
+                CharCount position = max(
+                    min(JavascriptRegExp::GetIndexOrMax(index), inputLength),
+                    (CharCount) 0);
+
+                // Capture groups can be referenced using at most two digits.
+                uint64 maxNumberOfCaptures = 99;
+                size_t numberOfCapturesToKeep = (size_t) min(numberOfCaptures, maxNumberOfCaptures);
+                if (captures == nullptr)
+                {
+                    captures = AnewArray(tempAlloc, Var, numberOfCapturesToKeep + 1);
+                }
+                else if (numberOfCapturesToKeep != previousNumberOfCapturesToKeep)
+                {
+                    size_t existingBytes = (previousNumberOfCapturesToKeep + 1) * sizeof(Var*);
+                    size_t requestedBytes = (numberOfCapturesToKeep + 1) * sizeof(Var*);
+                    captures = (Var*) tempAlloc->Realloc(captures, existingBytes, requestedBytes);
+                }
+                previousNumberOfCapturesToKeep = numberOfCapturesToKeep;
+
+                for (uint64 i = 1; i <= numberOfCaptures; ++i)
+                {
+                    Var nextCapture = JavascriptOperators::GetItem(resultObj, i, scriptContext);
+                    if (!JavascriptOperators::IsUndefined(nextCapture))
+                    {
+                        nextCapture = JavascriptConversion::ToString(nextCapture, scriptContext);
+                    }
+
+                    if (i <= numberOfCapturesToKeep)
+                    {
+                        captures[i] = nextCapture;
+                    }
```

Comment:
Yes, since `ToString` is observable, we need to run that on all captures. However, we don't need to keep a capture if it isn't possible to refer to it.